experimental/Schemes/Schemes.jl

```diff
@@ -0,0 +1,2587 @@
+  return D.pullbackFromRoot
+end
+
+@Markdown.doc """
+    pullback_from_root( X::Spec ) = AlgebraHomomorphism( X.R, X.R, gens(X.R) )
+
+Returns the identity of the ambient ring of X in order to be able to complete 
+inductive implementations on the localization tree conveniently.
+"""
+pullback_from_root( X::Spec ) = AlgebraHomomorphism( X.R, X.R, gens(X.R) )
+
+@Markdown.doc """
+    pullback_from_parent( X::Spec ) = AlgebraHomomorphism( X.R, X.R, gens(X.R) )
+
+Returns the identity of the ambient ring of X in order to be able to complete 
+inductive implementations on the localization tree conveniently.
+"""
+pullback_from_parent( X::Spec ) = AlgebraHomomorphism( X.R, X.R, gens(X.R) )
+
+@Markdown.doc """
+    function pullback_from_parent( D::SpecPrincipalOpen )
+    
+Returns the ring homomorphism of the ambient rings D' = parent(D) 
+and D lifting the pullback of functions for the inclusion D ↪ D'.
+"""
+function pullback_from_parent( D::SpecPrincipalOpen )
+  if isdefined( D, :pullbackFromParent )
+    return D.pullbackFromParent
+  end
+  ambient_ring( D ) # This also stores the homomorphism
+  return D.pullbackFromParent
+end
+
+@Markdown.doc """
+    function defining_ideal(D::SpecPrincipalOpen)
+
+Returns the ideal I + ⟨1-t⋅∏ᵢfᵢ⟩ of the ambient ring R[t] of D 
+where D = Spec (R/I)[(f₁)⁻¹,…,(fᵣ)⁻¹] with 
+R a free polynomial ring and t projecting to the element 1/∏ᵢfᵢ.
+"""
+function defining_ideal(D::SpecPrincipalOpen)
+  if isdefined( D, :I )
+    return D.I
+  end
+  ambient_ring(D) # This now also prepares the ideal.
+  return D.I
+end
+
+@Markdown.doc """
+    function inclusion_in_parent( D::SpecPrincipalOpen ) -> AffSchMorphism
+
+Return the inclusion D ↪ D' of a principal open subscheme D into its parent D'.
+"""
+function inclusion_in_parent( D::SpecPrincipalOpen )
+  # return AffSchMorphism( D, parent(D), pullback_from_parent(D) )
+  F = FractionField(ambient_ring(root(D)))
+  return AffSchMorphism( D, parent(D), [ F(x) for x in gens(ambient_ring(root(D))) ] )
+end
+
+@Markdown.doc """
+    function inclusion_in_parent( D::SpecPrincipalOpen ) -> AffSchMorphism
+
+Return the inclusion D ↪ X of a principal open subscheme D into its root X in 
+the localization tree.
+"""
+function inclusion_in_root( D::SpecPrincipalOpen )
+  # return AffSchMorphism( D, root(D), pullback_from_root(D) )
+  F = FractionField(ambient_ring(root(D)))
+  return AffSchMorphism( D, root(D), [ F(x) for x in gens(ambient_ring(root(D))) ] )
+end
+
+# outer constructors
+
+@Markdown.doc """
+    function Spec( k::S, R::T ) where{S <: Ring, T <:MPolyRing} -> Spec
+
+Returns the scheme X = Spec R with R an instance of MPolyRing over k.
+"""
+function Spec( k::S, R::T ) where{S <: Ring, T <:MPolyRing}
+  I = ideal(R, zero(R))
+  return Spec(k, R, I )
+end
+
+@doc Markdown.doc"""
+    Spec(R::MPolyRing) -> Spec
+
+Return the affine scheme corresponding to the ring $R/(0)$.
+"""
+function Spec( R::T ) where{T <: MPolyRing}
```

Comment:
When you don't use the type `T` in the body, and also don't need to assert a relation between the types of different arguments, you can simplify the code:
```suggestion
function Spec(R::MPolyRing)
```